rewrite-java/src/main/java/org/openrewrite/java/UseGettersAndSetters.java

```diff
@@ -0,0 +1,147 @@
+package org.openrewrite.java;
+
+import org.openrewrite.Formatting;
+import org.openrewrite.Tree;
+import org.openrewrite.Validated;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.TypeUtils;
+
+import java.util.List;
+
+import static org.openrewrite.internal.StringUtils.capitalize;
+
+public class UseGettersAndSetters extends JavaIsoRefactorVisitor {
+
+    private JavaType.Class classType;
+
+    public UseGettersAndSetters() {
+        setCursoringOn();
+    }
+
+    public void setClassType(String classType) {
+        this.classType = JavaType.Class.build(classType);
+    }
+
+    @Override
+    public Validated validate() {
+        return Validated.required("classType", classType);
+    }
+
+    @Override
+    public J.ClassDecl visitClassDecl(J.ClassDecl classDecl) {
+        if (classType != null && classDecl.getType() != null
+                && classType.getFullyQualifiedName().equals(classDecl.getType().getFullyQualifiedName())) {
+            andThen(new Scoped(classDecl));
+        }
+        return classDecl;
+    }
+
+    public static class Scoped extends JavaIsoRefactorVisitor {
+
+        private final J.ClassDecl scope;
+
+        public Scoped(J.ClassDecl scope) {
+            this.scope = scope;
+            setCursoringOn();
+        }
+
+        @Override
+        public J.ClassDecl visitClassDecl(J.ClassDecl classDecl) {
```

Comment:
@sambsnyd 

Is this the correct way to check the scope here? It's logically repetitive of visitClassDecl in the outer UseGettersAndSetters, but Scoped is public so that someone could use it directly if they have a J.ClassDecl to specify for the type to refactor instead of a String (scope here, classType in the outer class), and I'm certain limiting the refactoring to the target type should apply regardless of which visitor you use. I think it might be appropriate to duplicate the logic in the outer class the way I have it because that could avoid ever having to process the inner scoped visitor. Is that correct?

Reply:
@gadams00 yes this is correct way to check the scope.
RE: Limiting the refactoring to the target type. `isScope()` compares elements by their UUID. So it will only ever match exactly the J.ClassDecl that was passed in, even if it has gone through many rounds of transformation by other visitors. So `Scoped` will be limited to apply only to the target type in this way. It's just how that target type is determined is up to the caller of `Scoped`. This flexibility is intentional as it allows this to be used as a building block in situations where the criteria for its use are different or more complicated than what's in `UseGettersAndSetters`. Does that make sense? You could pass in the `JavaType.Class` but that wouldn't let you do anything you can't do with the J.ClassDecl (which you can just call getType() on if you want its `JavaType.Class` back.


Yes there is some redundancy/indirection in finding the AST element you want, then setting it aside for another visitor.
One alternative might be, instead of using `andThen()`, to directly call `visitClassDecl`, like so:
```java
// in the outer class UseGettersAndSetters
    @Override
    public J.ClassDecl visitClassDecl(J.ClassDecl classDecl) {
        J.ClassDecl cd = classDecl;
        if (classType != null && classDecl.getType() != null
                && classType.getFullyQualifiedName().equals(classDecl.getType().getFullyQualifiedName())) {
            cd  = new Scoped(classDecl).visitClassDecl(classDecl);
        }
        return cd;
    }
```

This avoids the indirection and should work just fine in this case. But some visitors need to visit everything, starting from the compilation unit, in order to work properly. Most visitors which need to use cursoring are in this category. So I consider this approach a performance optimization rather than the default way of doing it. 

Does that answer your question? I wrote a lot which is sometimes I sign I haven't fully understood the question.

Reply:
Yeah, it does answer the question. I was just instinctively concerned by the redundancy, and wanted to validate my thinking that is was fine. Thanks!